pygit2 installation blocker: "fatal error: 'git2.h' file not found"

I’m helping someone set up GFtools on their computer to build a font, and we’ve hit an unexpected snag.

Gftools won’t install in the venv of a project folder, which they got from a repo on GitHub via the "Download Zip" method. The installation is blocked by pygit2 with the error shown below.

<details>
<summary><b><!-------->Installation log for pygit2<!--------></b> (Click to expand)</summary>

```
Building wheels for collected packages: pygit2
  Building wheel for pygit2 (pyproject.toml) ... error
  error: subprocess-exited-with-error
  
  × Building wheel for pygit2 (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [69 lines of output]
      running bdist_wheel
      running build
      running build_py
      creating build/lib.macosx-14.0-arm64-cpython-313/pygit2
      copying pygit2/enums.py -> build/lib.macosx-14.0-arm64-cpython-313/pygit2
      copying pygit2/legacyenums.py -> build/lib.macosx-14.0-arm64-cpython-313/pygit2
      copying pygit2/config.py -> build/lib.macosx-14.0-arm64-cpython-313/pygit2
      copying pygit2/branches.py -> build/lib.macosx-14.0-arm64-cpython-313/pygit2
      copying pygit2/submodules.py -> build/lib.macosx-14.0-arm64-cpython-313/pygit2
      copying pygit2/index.py -> build/lib.macosx-14.0-arm64-cpython-313/pygit2
      copying pygit2/remotes.py -> build/lib.macosx-14.0-arm64-cpython-313/pygit2
      copying pygit2/references.py -> build/lib.macosx-14.0-arm64-cpython-313/pygit2
      copying pygit2/packbuilder.py -> build/lib.macosx-14.0-arm64-cpython-313/pygit2
      copying pygit2/credentials.py -> build/lib.macosx-14.0-arm64-cpython-313/pygit2
      copying pygit2/__init__.py -> build/lib.macosx-14.0-arm64-cpython-313/pygit2
      copying pygit2/_build.py -> build/lib.macosx-14.0-arm64-cpython-313/pygit2
      copying pygit2/refspec.py -> build/lib.macosx-14.0-arm64-cpython-313/pygit2
      copying pygit2/utils.py -> build/lib.macosx-14.0-arm64-cpython-313/pygit2
      copying pygit2/blame.py -> build/lib.macosx-14.0-arm64-cpython-313/pygit2
      copying pygit2/callbacks.py -> build/lib.macosx-14.0-arm64-cpython-313/pygit2
      copying pygit2/settings.py -> build/lib.macosx-14.0-arm64-cpython-313/pygit2
      copying pygit2/filter.py -> build/lib.macosx-14.0-arm64-cpython-313/pygit2
      copying pygit2/errors.py -> build/lib.macosx-14.0-arm64-cpython-313/pygit2
      copying pygit2/_run.py -> build/lib.macosx-14.0-arm64-cpython-313/pygit2
      copying pygit2/repository.py -> build/lib.macosx-14.0-arm64-cpython-313/pygit2
      copying pygit2/ffi.py -> build/lib.macosx-14.0-arm64-cpython-313/pygit2
      copying pygit2/blob.py -> build/lib.macosx-14.0-arm64-cpython-313/pygit2
      copying pygit2/_pygit2.pyi -> build/lib.macosx-14.0-arm64-cpython-313/pygit2
      creating build/lib.macosx-14.0-arm64-cpython-313/pygit2/decl
      copying pygit2/decl/oid.h -> build/lib.macosx-14.0-arm64-cpython-313/pygit2/decl
      copying pygit2/decl/index.h -> build/lib.macosx-14.0-arm64-cpython-313/pygit2/decl
      copying pygit2/decl/attr.h -> build/lib.macosx-14.0-arm64-cpython-313/pygit2/decl
      copying pygit2/decl/blame.h -> build/lib.macosx-14.0-arm64-cpython-313/pygit2/decl
      copying pygit2/decl/pack.h -> build/lib.macosx-14.0-arm64-cpython-313/pygit2/decl
      copying pygit2/decl/revert.h -> build/lib.macosx-14.0-arm64-cpython-313/pygit2/decl
      copying pygit2/decl/net.h -> build/lib.macosx-14.0-arm64-cpython-313/pygit2/decl
      copying pygit2/decl/config.h -> build/lib.macosx-14.0-arm64-cpython-313/pygit2/decl
      copying pygit2/decl/types.h -> build/lib.macosx-14.0-arm64-cpython-313/pygit2/decl
      copying pygit2/decl/repository.h -> build/lib.macosx-14.0-arm64-cpython-313/pygit2/decl
      copying pygit2/decl/clone.h -> build/lib.macosx-14.0-arm64-cpython-313/pygit2/decl
      copying pygit2/decl/callbacks.h -> build/lib.macosx-14.0-arm64-cpython-313/pygit2/decl
      copying pygit2/decl/submodule.h -> build/lib.macosx-14.0-arm64-cpython-313/pygit2/decl
      copying pygit2/decl/errors.h -> build/lib.macosx-14.0-arm64-cpython-313/pygit2/decl
      copying pygit2/decl/merge.h -> build/lib.macosx-14.0-arm64-cpython-313/pygit2/decl
      copying pygit2/decl/graph.h -> build/lib.macosx-14.0-arm64-cpython-313/pygit2/decl
      copying pygit2/decl/describe.h -> build/lib.macosx-14.0-arm64-cpython-313/pygit2/decl
      copying pygit2/decl/stash.h -> build/lib.macosx-14.0-arm64-cpython-313/pygit2/decl
      copying pygit2/decl/remote.h -> build/lib.macosx-14.0-arm64-cpython-313/pygit2/decl
      copying pygit2/decl/strarray.h -> build/lib.macosx-14.0-arm64-cpython-313/pygit2/decl
      copying pygit2/decl/buffer.h -> build/lib.macosx-14.0-arm64-cpython-313/pygit2/decl
      copying pygit2/decl/diff.h -> build/lib.macosx-14.0-arm64-cpython-313/pygit2/decl
      copying pygit2/decl/common.h -> build/lib.macosx-14.0-arm64-cpython-313/pygit2/decl
      copying pygit2/decl/proxy.h -> build/lib.macosx-14.0-arm64-cpython-313/pygit2/decl
      copying pygit2/decl/indexer.h -> build/lib.macosx-14.0-arm64-cpython-313/pygit2/decl
      copying pygit2/decl/refspec.h -> build/lib.macosx-14.0-arm64-cpython-313/pygit2/decl
      copying pygit2/decl/transport.h -> build/lib.macosx-14.0-arm64-cpython-313/pygit2/decl
      copying pygit2/decl/checkout.h -> build/lib.macosx-14.0-arm64-cpython-313/pygit2/decl
      copying pygit2/decl/commit.h -> build/lib.macosx-14.0-arm64-cpython-313/pygit2/decl
      running build_ext
      generating cffi module 'build/temp.macosx-14.0-arm64-cpython-313/pygit2._libgit2.c'
      creating build/temp.macosx-14.0-arm64-cpython-313
      building 'pygit2._pygit2' extension
      creating build/temp.macosx-14.0-arm64-cpython-313/private/var/folders/yh/bfs3pr6n58q5wx6kpfwk8h7r0000gn/T/pip-install-9c_y2296/pygit2_9689cf534fe54ad9960ea3b3d516eb1c/src
      clang -fno-strict-overflow -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -O3 -Wall -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk -I/usr/local/include -I/Users/username/Dropbox/projectname/venv/include -I/opt/homebrew/opt/python@3.13/Frameworks/Python.framework/Versions/3.13/include/python3.13 -c /private/var/folders/yh/bfs3pr6n58q5wx6kpfwk8h7r0000gn/T/pip-install-9c_y2296/pygit2_9689cf534fe54ad9960ea3b3d516eb1c/src/blob.c -o build/temp.macosx-14.0-arm64-cpython-313/private/var/folders/yh/bfs3pr6n58q5wx6kpfwk8h7r0000gn/T/pip-install-9c_y2296/pygit2_9689cf534fe54ad9960ea3b3d516eb1c/src/blob.o
      /private/var/folders/yh/bfs3pr6n58q5wx6kpfwk8h7r0000gn/T/pip-install-9c_y2296/pygit2_9689cf534fe54ad9960ea3b3d516eb1c/src/blob.c:30:10: fatal error: 'git2.h' file not found
         30 | #include <git2.h>
            |          ^~~~~~~~
      1 error generated.
      error: command '/usr/bin/clang' failed with exit code 1
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for pygit2
Failed to build pygit2
ERROR: ERROR: Failed to build installable wheels for some pyproject.toml based projects (pygit2)
```

</details>

At first, the problem/solution seemed obvious to me: maybe GFtools / pygit2 can _only_ be installed in a Git repo. The weird thing is, _it is_ installing in another project folder, and this project folder was also simply downloaded as a zip from GitHub, and also contains no `.git` directory.

I had hoped that Simon’s notes at https://github.com/googlefonts/gftools/issues/442 might help, but `brew install libgit2; LIBGIT2=/opt/homebrew pip3 install pygit2` didn’t seem to work to get the installation going.

We specifically tried items like `pip install -U gftools` and `pip install --no-cache-dir gftools` and `pip install gftools==0.9.71`, but these all gave the same installation error.

Does anyone have any idea of what might be going wrong? I don’t quite know what to test next. 

(One thought... Could it be a Python version issue? I believe the person trying to install this is on Python that is a couple of versions back... I’ll ask them to update and try again. But yeah, short of that, I’m not sure what to try next!)

Thanks so much for any ideas here!

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -36,7 +36,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        python-version: ["3.8", "3.9", "3.10", "3.11"]
+        python-version: ["3.9", "3.10", "3.11", "3.13"]
         platform: [ubuntu-latest, windows-latest]
     steps:
       - uses: actions/checkout@v4

--- a/Lib/gftools/packager/__init__.py
+++ b/Lib/gftools/packager/__init__.py
@@ -423,7 +423,7 @@ def git_tree_traverse(func, *args, **kwargs):
         tree = repo.get(tree_oid)
         try:
             entry = tree[subtree_name]
-            existing_subtree = repo.get(entry.hex)
+            existing_subtree = repo.get(entry.id)
             sub_treebuilder = repo.TreeBuilder(existing_subtree)
         except KeyError:
             sub_treebuilder = repo.TreeBuilder()

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -44,7 +44,9 @@ dependencies = [
   'unidecode',
   'opentype-sanitizer',
   'vttlib',
-  'pygit2==1.16.0',
+  # pygit2 1.16.0 only supports 3.10+
+  'pygit2==1.15.0; python_version < "3.13"',
+  'pygit2==1.16.0; python_version >= "3.13"',
   'strictyaml',
   'fontmake[json]>=3.3.0',
   'skia-pathops',

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -44,7 +44,7 @@ dependencies = [
   'unidecode',
   'opentype-sanitizer',
   'vttlib',
-  'pygit2<=1.14.1',
+  'pygit2==1.16.0',
   'strictyaml',
   'fontmake[json]>=3.3.0',
   'skia-pathops',

